numba/np/arraymath.py

```diff
@@ -669,6 +670,131 @@ def _trivial_isnan(x):
+    if type_can_asarray(x):
+        return lambda x: _impl_iscomplex_isreal_arr(operator_fn, x)
+    elif isinstance(x, types.Optional):
+        return None
+    elif is_nonelike(x):
+        def impl(x):
+            return opt_ret
+    else:
+        def impl(x):
+            return False
+    return impl
+
+
+@overload(np.iscomplex)
+def np_iscomplex(x):
+    return np_iscomplex_isreal(x, operator.ne, False)
+
+
+@overload(np.isreal)
+def np_isreal(x):
+    return np_iscomplex_isreal(x, operator.eq, True)
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    if isinstance(x, types.Optional):
+        msg = 'Expected argument {} but got Optional[{}, None]'.format(x.key, x.key)
+        raise TypingError(msg)
+    dt = determine_dtype(x)
+    iscmplx = np.issubdtype(dt, np.complexfloating)
+    return lambda x: False if x is None else iscmplx
+
+
+@overload(np.isrealobj)
+def isrealobj(x):
+    # Return True if x is a not complex type or an array of complex numbers.
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/lib/type_check.py#L290-L322
+    def impl(x):
+        return not np.iscomplexobj(x)
+    return impl
+
+
+@overload(np.isscalar)
+def np_isscalar(num):
+    res = isinstance(num, (types.Number, types.UnicodeType, types.Boolean))
+
+    def impl(num):
+        return res
+    return impl
+
+
+@register_jitable
+def _within_tol(a, b, rtol, atol):
+    return np.less_equal(np.abs(a - b), atol + rtol * np.abs(b))
+
+
+@overload(np.isclose)
+def np_isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+    # Based on NumPy impl.
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/core/numeric.py#L2180-L2292
+
+    if not(type_can_asarray(a) and type_can_asarray(b)):
+        raise TypingError("`np.isclose` is not supported for the input types.")
+
+    def impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+        xfin = np.asarray(np.isfinite(a))
+        yfin = np.asarray(np.isfinite(b))
+        if np.all(xfin) and np.all(yfin):
+            return _within_tol(a, b, rtol, atol)
+        else:
+            r = _within_tol(a, b, rtol, atol)
+            if equal_nan:
+                return r | (~xfin & ~yfin)
+            else:
+                return r
+    return impl
+
+
+def is_np_inf_impl(x, out, fn):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    if is_nonelike(out):
+        if isinstance(x, (types.Array, types.Sequence, types.Tuple)):
+            def impl(x, out=None):
+                x = np.asarray(x)
+                out = np.zeros(x.shape, dtype=types.boolean)
+                np.logical_and(np.isinf(x), fn(np.signbit(x)), out)
+                return out
```

Comment:
Is this worthwhile vs the simpler:
```suggestion
                return np.logical_and(np.isinf(x), fn(np.signbit(x)), out)
```

It seems to me that `np.logical_and` would contain exactly this logic internally anyway

Reply:
Great suggestion